pkg/kvstore/allocator.go

```diff
@@ -0,0 +1,539 @@
+//  - WithEvents() - enable Events channel
+//  - WithSuffix(string) - customize the node specifix suffix to attach to keys
+//  - WithMin(id) - minimum ID to allocate (default: 1)
+//  - WithMax(id) - maximum ID to allocate (default max(uint64))
+//
+// After creation, IDs can be allocated with Allocate() and released with
+// Release()
+func NewAllocator(basePath string, typ AllocatorKey, opts ...AllocatorOption) (*Allocator, error) {
+	if Client == nil {
+		return nil, fmt.Errorf("kvstore client not configured")
+	}
+
+	a := &Allocator{
+		keyType:     typ,
+		idPrefix:    basePath + "/id/",
+		valuePrefix: basePath + "/value/",
+		lockPrefix:  basePath + "/",
+		min:         1,
+		max:         ID(^uint64(0)),
+		localKeys:   newLocalKeys(),
+		stopGC:      make(chan bool, 0), // unbuffered channel so gc is stopped in sync
+		suffix:      nodeaddress.GetExternalIPv4().String(),
+	}
+
+	for _, fn := range opts {
+		fn(a)
+	}
+
+	a.startGC()
+
+	return a, nil
+}
+
+// WithEvents enables AllocatorEvent to be received on Allocator.Events
+func WithEvents() AllocatorOption {
+	return func(a *Allocator) { a.startWatch() }
+}
+
+// WithSuffix sets the suffix of the allocator to the specified value
+func WithSuffix(v string) AllocatorOption {
+	return func(a *Allocator) { a.suffix = v }
+}
+
+// WithMin sets the minimum identifier to be allocated
+func WithMin(id ID) AllocatorOption {
+	return func(a *Allocator) { a.min = id }
+}
+
+// WithMax sets the maximum identifier to be allocated
+func WithMax(id ID) AllocatorOption {
+	return func(a *Allocator) { a.max = id }
+}
+
+// Delete deletes an allocator and stops the garbage collector
+func (a *Allocator) Delete() {
+	a.stopGC <- true
+	a.stopWatch()
+}
+
+// DeleteAllKeys will delete all keys
+func (a *Allocator) DeleteAllKeys() {
+	Client.DeleteTree(a.idPrefix)
+	Client.DeleteTree(a.valuePrefix)
+	Client.DeleteTree(a.lockPrefix)
+}
+
+func invalidKey(key, prefix string, deleteInvalid bool) {
+	log.Warningf("kvstore: Found invalid key %s outside of prefix %s. Deleting...", key, prefix)
+
+	if deleteInvalid {
+		Delete(key)
+	}
+}
+
+func (a *Allocator) keyToID(key string, deleteInvalid bool) ID {
+	if !strings.HasPrefix(key, a.idPrefix) {
+		invalidKey(key, a.idPrefix, deleteInvalid)
+		return 0
+	}
+
+	id, _ := strconv.ParseUint(key[len(a.idPrefix):], 10, 64)
+	return ID(id)
+}
+
+// Naive ID allocation mechanism.
+// FIXME: This should consider a random ID in the future
+func (a *Allocator) selectAvailableID() (ID, string) {
+	for id := a.min; id <= a.max; id++ {
+		if _, ok := a.cache[id]; !ok {
+			return id, id.String()
+		}
+	}
+
+	return 0, ""
+}
+
+// RefillCache invalidates and refills the local cache. Afterwards,
+// Allocator.Cache will contain entries for all IDs to key mappings as found in
+// the KVstore.
+func (a *Allocator) RefillCache() error {
+	// fetch all /id/ keys
+	ids, err := ListPrefix(a.idPrefix)
+	if err != nil {
+		return err
+	}
+
+	newCache := IDMap{}
```

Comment:
You can allocate a map of the right size right away:

    newCache := IDMap(make(map[ID]AllocatorKey, len(ids)))